File: extension/unidirectional_lstm.cc
#include "extension/unidirectional_lstm.h"
#include <cmath>
#include <stdexcept>

namespace cnt {

void block_orthogonal_recursive(
    const at::IntList sizes,
    const at::IntList split_sizes,
    const double gain,
    const int64_t dim,
    torch::Tensor pre_narrowed_tensor) {
  if (dim == static_cast<int64_t>(split_sizes.size())) {
    // Find a block.
    torch::nn::init::orthogonal_(pre_narrowed_tensor, gain);
    return;
  }

  for (int64_t start = 0; start < sizes[dim]; start += split_sizes[dim]) {
    block_orthogonal_recursive(
        sizes,
        split_sizes,
        gain,
        dim + 1,
        pre_narrowed_tensor.narrow(dim, start, split_sizes[dim]));
  }
}

void block_orthogonal(
    torch::Tensor tensor,
    at::IntList split_sizes,
    double gain = 1.0) {
  // Get the sizes of ``tensor``.
  auto sizes = tensor.sizes();

  // Check ``split_sizes``:
  if (split_sizes.size() != sizes.size()) {
    throw std::invalid_argument("Dimension not match: split_sizes, sizes");
  }
  for (int64_t dim = 0; dim < static_cast<int64_t>(split_sizes.size()); dim++) {
    if (sizes[dim] % split_sizes[dim] != 0) {
      throw std::invalid_argument("Size divisible: split_sizes, sizes");
    }
  }

  // Apply ``torch::nn::init::orthogonal_`` to all combinations of slices.
  block_orthogonal_recursive(
      sizes,
      split_sizes,
      gain,
      0,
      tensor);
}

torch::Tensor get_dropout_mask(
    double dropout_probability,
    torch::Tensor tensor_for_masking) {
  auto binary_mask = torch::rand(tensor_for_masking.sizes())
      .gt(dropout_probability)
      .to(tensor_for_masking.device());
  auto dropout_mask = binary_mask
      .to(torch::dtype(torch::kFloat32))
      .div(1.0 - dropout_probability);
  return dropout_mask;
}

UnidirectionalSingleLayerLstmImpl::UnidirectionalSingleLayerLstmImpl(
    int64_t input_size,
    int64_t hidden_size,
    int64_t cell_size,
    bool go_forward,
    int64_t truncated_bptt,
    double cell_clip,
    double proj_clip,
    int64_t recurrent_dropout_type,
    double recurrent_dropout_probability)
    :
    input_size_(input_size),
    hidden_size_(hidden_size),
    cell_size_(cell_size),
    go_forward_(go_forward),
    truncated_bptt_(truncated_bptt),
    cell_clip_(fabs(cell_clip)),
    proj_clip_(fabs(proj_clip)),
    recurrent_dropout_type_(recurrent_dropout_type),
    recurrent_dropout_probability_(recurrent_dropout_probability) {
  // Create & bind linear projections.
  input_linearity_weight_ = register_parameter(
      "input_linearity_weight",
      torch::zeros(
          {4 * cell_size_, input_size_},
          torch::dtype(torch::kFloat32)));

  hidden_linearity_weight_ = register_parameter(
      "hidden_linearity_weight",
      torch::zeros(
          {4 * cell_size_, hidden_size_},
          torch::dtype(torch::kFloat32)));
  hidden_linearity_bias_ = register_parameter(
      "hidden_linearity_bias",
      torch::zeros(
          {4 * cell_size_},
          torch::dtype(torch::kFloat32)));

  proj_linearity_weight_ = register_parameter(
      "proj_linearity_weight",
      torch::zeros(
          {hidden_size_, cell_size_},
          torch::dtype(torch::kFloat32)));

  // Reset parameters.
  block_orthogonal(
      input_linearity_weight_.detach(),
      {cell_size_, input_size_});

  block_orthogonal(
      hidden_linearity_weight_.detach(),
      {cell_size_, hidden_size_});
  hidden_linearity_bias_
      .detach()
      .fill_(0.0);
  hidden_linearity_bias_
      .detach()
      .narrow(0, cell_size_, cell_size_).fill_(1.0);

  block_orthogonal(
      proj_linearity_weight_.detach(),
      {hidden_size_, cell_size_});
}

LstmForwardRetType UnidirectionalSingleLayerLstmImpl::forward(
    torch::Tensor inputs,
    torch::Tensor batch_sizes,
    LstmStateType initial_state) {
  // Reduce the cost of dynamic dispatch.
  auto batch_sizes_accessor = batch_sizes.accessor<int64_t, 1>();

  int64_t total_sequences_size = inputs.size(0);
  int64_t total_timesteps = batch_sizes_accessor.size(0);

  // Unpack hidden/cell state.
  auto initial_hidden_state = std::get<0>(initial_state).squeeze(0);
  auto initial_cell_state = std::get<1>(initial_state).squeeze(0);
  // `batch_sizes_accessor[0]`: the batch size.
  if (initial_hidden_state.size(0) < batch_sizes_accessor[0]
      || initial_cell_state.size(0) < batch_sizes_accessor[0]) {
    throw std::invalid_argument(
        "initial_state can't fullfill inputs.");
  }

  // Initialize the output container.
  std::vector<torch::Tensor> output_accumulator(total_timesteps);
  // Keep the chunks of hidden/cell in forward mode.
  std::vector<torch::Tensor> forward_hiddens;
  std::vector<torch::Tensor> forward_cells;

  // Create aliases for dropout.
  auto batch_input_linearity_weight = input_linearity_weight_;
  auto batch_hidden_linearity_weight = hidden_linearity_weight_;

  // Variational dropout mask on hidden state.
  torch::Tensor variational_dropout_mask;
  if (is_training() && recurrent_dropout_type_ == 1) {
    if (recurrent_dropout_probability_ == 0.0) {
      throw std::invalid_argument(
          "Variational: recurrent_dropout_probability not provided.");
    }
    variational_dropout_mask = get_dropout_mask(
        recurrent_dropout_probability_,
        initial_hidden_state);
  }

  // Dropconnect mask on input/hidden linear projection weights.
  torch::Tensor dropconnect_input_mask;
  torch::Tensor dropconnect_hidden_mask;
  if (is_training() && recurrent_dropout_type_ == 2) {
    if (recurrent_dropout_probability_ == 0.0) {
      throw std::invalid_argument(
          "Dropconnect: recurrent_dropout_probability not provided.");
    }
    dropconnect_input_mask = get_dropout_mask(
        recurrent_dropout_probability_,
        input_linearity_weight_);
    dropconnect_hidden_mask = get_dropout_mask(
        recurrent_dropout_probability_,
        hidden_linearity_weight_);
  }
  // Apply Dropconnect.
  if (is_training() && recurrent_dropout_type_ == 2) {
    batch_input_linearity_weight =
        input_linearity_weight_ * dropconnect_input_mask;
    batch_hidden_linearity_weight =
        hidden_linearity_weight_ * dropconnect_hidden_mask;
  }

  // Temporal hidden/cell.
  int64_t init_batch_size =
      go_forward_ ?
      batch_sizes_accessor[0] :
      batch_sizes_accessor[total_timesteps - 1];
  auto hidden_state = initial_hidden_state.narrow(0, 0, init_batch_size);
  auto cell_state = initial_cell_state.narrow(0, 0, init_batch_size);

  // Loop over each timestep and unpack inputs manually.
  int64_t total_offset = 0;
  for (int64_t ts_offset = 0; ts_offset < total_timesteps; ts_offset++) {
    int64_t timestep_index =
        go_forward_ ? ts_offset : total_timesteps - 1 - ts_offset;
    int64_t batch_size = batch_sizes_accessor[timestep_index];

    // TF-style TBPTT.
    // See https://r2rt.com/styles-of-truncated-backpropagation.html.
    if (is_training()
          && truncated_bptt_ > 0
          && ts_offset > 0
          && ts_offset % truncated_bptt_ == 0) {
      hidden_state = hidden_state.detach();
      cell_state = cell_state.detach();
    }

    // Slicing hidden/cell.
    if (go_forward_) {
      if (timestep_index > 0) {
        // Trim off if necessary.
        auto last_batch_size = batch_sizes_accessor[timestep_index - 1];
        auto dec = last_batch_size - batch_size;
        if (dec > 0) {
          forward_hiddens.push_back(
              hidden_state.narrow(
                  0,
                  last_batch_size - dec,
                  dec));
          forward_cells.push_back(
              cell_state.narrow(
                  0,
                  last_batch_size - dec,
                  dec));
          // Narrows state.
          hidden_state = hidden_state.narrow(0, 0, batch_size);
          cell_state = cell_state.narrow(0, 0, batch_size);
        }
      }
    } else {
      if (timestep_index < total_timesteps - 1) {
        auto last_batch_size = batch_sizes_accessor[timestep_index + 1];
        auto inc = batch_size - last_batch_size;
        if (inc > 0) {
          hidden_state = torch::cat(
              {
                  hidden_state,
                  initial_hidden_state.narrow(0, batch_size - inc, inc),
              },
              0);
          cell_state = torch::cat(
              {
                  cell_state,
                  initial_cell_state.narrow(0, batch_size - inc, inc),
              },
              0);
        }
      }
    }

    // Slicing inputs.
    int64_t input_slice_begin = -1;
    if (go_forward_) {
      input_slice_begin = total_offset;
    } else {
      input_slice_begin = total_sequences_size - total_offset - batch_size;
    }
    auto cur_input = inputs.narrow(0, input_slice_begin, batch_size);

    // Compute next hidden/cell state.
    auto proj_input = torch::linear(
        cur_input, batch_input_linearity_weight);
    auto proj_hidden = torch::linear(
        hidden_state, batch_hidden_linearity_weight, hidden_linearity_bias_);

    auto input_gate = torch::sigmoid(
        proj_input.narrow(1, 0 * cell_size_, cell_size_) +
        proj_hidden.narrow(1, 0 * cell_size_, cell_size_));
    auto forget_gate = torch::sigmoid(
        proj_input.narrow(1, 1 * cell_size_, cell_size_) +
        proj_hidden.narrow(1, 1 * cell_size_, cell_size_));
    auto cell_tilde = torch::tanh(
        proj_input.narrow(1, 2 * cell_size_, cell_size_) +
        proj_hidden.narrow(1, 2 * cell_size_, cell_size_));
    auto output_gate = torch::sigmoid(
        proj_input.narrow(1, 3 * cell_size_, cell_size_) +
        proj_hidden.narrow(1, 3 * cell_size_, cell_size_));

    cell_state = input_gate * cell_tilde + forget_gate * cell_state;
    if (cell_clip_ > 0.0) {
      cell_state.clamp_(-cell_clip_, cell_clip_);
    }

    hidden_state = torch::linear(
        output_gate * torch::tanh(cell_state),
        proj_linearity_weight_);
    if (proj_clip_ > 0.0) {
      hidden_state.clamp_(-proj_clip_, proj_clip_);
    }

    // Apply variational dropout.
    if (is_training() && recurrent_dropout_type_ == 1) {
      hidden_state *=
          variational_dropout_mask.narrow(0, 0, batch_size);
    }

    // Update offset.
    total_offset += batch_size;

    // Fill hidden state to output.
    output_accumulator[timestep_index] = hidden_state;
  }

  if (go_forward_) {
    forward_hiddens.push_back(hidden_state);
    std::reverse(forward_hiddens.begin(), forward_hiddens.end());
    hidden_state = torch::cat(forward_hiddens, 0);

    forward_cells.push_back(cell_state);
    std::reverse(forward_cells.begin(), forward_cells.end());
    cell_state = torch::cat(forward_cells, 0);
  }

  auto state_dec = initial_hidden_state.size(0) - hidden_state.size(0);
  if (state_dec > 0) {
    hidden_state = torch::cat(
        {
            hidden_state,
            initial_hidden_state.narrow(0, hidden_state.size(0), state_dec),
        },
        0);
    cell_state = torch::cat(
        {
            cell_state,
            initial_cell_state.narrow(0, cell_state.size(0), state_dec),
        },
        0);
  }

  return std::make_tuple(
      torch::cat(output_accumulator, 0),
      std::make_tuple(
          hidden_state.unsqueeze(0),
          cell_state.unsqueeze(0)));
}

torch::TensorOptions UnidirectionalSingleLayerLstmImpl::weight_options() {
  return torch::dtype(torch::kFloat32)
      .device(hidden_linearity_weight_.device());
}

UnidirectionalLstmImpl::UnidirectionalLstmImpl(
    int64_t num_layers,
    int64_t input_size,
    int64_t hidden_size,
    int64_t cell_size,
    bool go_forward,
    int64_t truncated_bptt,
    double cell_clip,
    double proj_clip,
    int64_t recurrent_dropout_type,
    double recurrent_dropout_probability,
    bool use_skip_connections)
    :
    hidden_size_(hidden_size),
    cell_size_(cell_size),
    num_layers_(num_layers),
    use_skip_connections_(use_skip_connections),
    layer_name_prefix_(go_forward ? "forward_layer_" : "backward_layer_") {
  if (num_layers_ == 0) {
    throw std::invalid_argument("num_layers should >= 1.");
  }

  auto lstm_input_size = input_size;

  for (int64_t layer_idx = 0; layer_idx < num_layers_; layer_idx++) {
    // Construct.
    auto layer = UnidirectionalSingleLayerLstm(
        lstm_input_size,
        hidden_size,
        cell_size,
        go_forward,
        truncated_bptt,
        cell_clip,
        proj_clip,
        recurrent_dropout_type,
        recurrent_dropout_probability);
    // Register.
    register_module(
        layer_name_prefix_ + std::to_string(layer_idx),
        layer);
    layers_.push_back(layer);

    lstm_input_size = hidden_size;
  }
}

LstmForwardMultiLayerRetType UnidirectionalLstmImpl::forward(
    torch::Tensor inputs,
    torch::Tensor batch_sizes,
    LstmStateType initial_state) {
  // Unpack initial_state.
  auto layers_hidden_state = std::get<0>(initial_state);
  auto layers_cell_state = std::get<1>(initial_state);

  std::vector<torch::Tensor> output_accumulators(num_layers_);
  std::vector<torch::Tensor> hidden_states(num_layers_);
  std::vector<torch::Tensor> cell_states(num_layers_);

  auto layer_inputs = inputs;
  for (int64_t layer_idx = 0; layer_idx < num_layers_; layer_idx++) {
    auto layer_hidden_state = layers_hidden_state
        .select(0, layer_idx).unsqueeze(0);
    auto layer_cell_state = layers_cell_state
        .select(0, layer_idx).unsqueeze(0);

    auto layer_out = layers_[layer_idx](
        layer_inputs,
        batch_sizes,
        std::make_tuple(
            layer_hidden_state,
            layer_cell_state));

    auto output_accumulator = std::get<0>(layer_out);
    auto final_state = std::get<1>(layer_out);
    auto hidden_state = std::get<0>(final_state);
    auto cell_state = std::get<1>(final_state);

    output_accumulators[layer_idx] = output_accumulator;
    hidden_states[layer_idx] = hidden_state;
    cell_states[layer_idx] = cell_state;

    if (layer_idx > 0 && use_skip_connections_) {
      // input_size == hidden_size.
      output_accumulator += layer_inputs;
    }

    layer_inputs = output_accumulator;
  }

  return std::make_tuple(
      output_accumulators,
      std::make_tuple(
          torch::cat(hidden_states, 0),
          torch::cat(cell_states, 0)));
}

torch::TensorOptions UnidirectionalLstmImpl::weight_options() {
  return layers_[0]->weight_options();
}

}  // namespace cnt

File: pytorch_stateful_lstm/unidirectional_lstm.py
"""
Python implementation of unidirectional_lstm.cc.
"""
from typing import Tuple, List
import math

import torch
from torch.nn import functional as F
from torch.nn.parameter import Parameter


def block_orthogonal_recursive(
        sizes: torch.Size,
        split_sizes: List[int],
        gain: float,
        dim: int,
        tensor: torch.Tensor,
) -> None:
    if dim == tensor.dim():
        torch.nn.init.orthogonal_(tensor, gain)
        return

    for start in range(0, sizes[dim], split_sizes[dim]):
        block_orthogonal_recursive(
                sizes,
                split_sizes,
                gain,
                dim + 1,
                tensor.narrow(dim, start, split_sizes[dim]),
        )


def block_orthogonal(
        tensor: torch.Tensor,
        split_sizes: List[int],
        gain: float = 1.0,
) -> None:
    sizes = tensor.shape

    if len(split_sizes) != len(sizes):
        raise ValueError("Dimension not match: split_sizes, sizes")

    for size_dim, split_size_dim in zip(sizes, split_sizes):
        if size_dim % split_size_dim != 0:
            raise ValueError("Size divisible: split_sizes, sizes")

    block_orthogonal_recursive(
            sizes,
            split_sizes,
            gain,
            0,
            tensor,
    )


def get_dropout_mask(
        dropout_probability: float,
        tensor_for_masking: torch.Tensor,
) -> torch.Tensor:
    binary_mask = torch.rand_like(tensor_for_masking).gt(dropout_probability)
    dropout_mask = binary_mask.to(torch.float).div(1.0 - dropout_probability)
    return dropout_mask


class PyUnidirectionalSingleLayerLstm(torch.nn.Module):  # type: ignore

    def __init__(
            self,
            input_size: int,
            hidden_size: int,
            cell_size: int,
            go_forward: bool = True,
            truncated_bptt: int = 0,
            cell_clip: float = 0.0,
            proj_clip: float = 0.0,
            recurrent_dropout_type: int = 0,
            recurrent_dropout_probability: float = 0.0,
    ) -> None:
        super().__init__()

        self.input_size = input_size
        self.hidden_size = hidden_size
        self.cell_size = cell_size
        self.go_forward = go_forward
        self.truncated_bptt = truncated_bptt
        self.cell_clip = math.fabs(cell_clip)
        self.proj_clip = math.fabs(proj_clip)
        self.recurrent_dropout_type = recurrent_dropout_type
        self.recurrent_dropout_probability = recurrent_dropout_probability

        self.input_linearity_weight = Parameter(torch.zeros((4 * cell_size, input_size)))
        block_orthogonal(self.input_linearity_weight.detach(), [cell_size, input_size])

        self.hidden_linearity_weight = Parameter(torch.zeros((4 * cell_size, hidden_size)))
        block_orthogonal(self.hidden_linearity_weight.detach(), [cell_size, hidden_size])

        self.hidden_linearity_bias = Parameter(torch.zeros((4 * cell_size,)))
        self.hidden_linearity_bias.detach().narrow(0, cell_size, cell_size).fill_(1.0)

        self.proj_linearity_weight = Parameter(torch.zeros((hidden_size, cell_size)))
        block_orthogonal(self.proj_linearity_weight.detach(), [hidden_size, cell_size])

    def forward(  # pylint: disable=arguments-differ
            self,
            inputs: torch.Tensor,
            batch_sizes: torch.Tensor,
            initial_state: Tuple[torch.Tensor, torch.Tensor],
    ) -> Tuple[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
        total_sequences_size = inputs.shape[0]
        total_timesteps = batch_sizes.shape[0]

        initial_hidden_state, initial_cell_state = initial_state
        initial_hidden_state = initial_hidden_state.squeeze(0)
        initial_cell_state = initial_cell_state.squeeze(0)

        if initial_hidden_state.shape[0] < batch_sizes[0] \
                or initial_cell_state.shape[0] < batch_sizes[0]:
            raise ValueError("initial_state can't fullfill inputs.")

        output_accumulator = [None] * total_timesteps
        forward_hiddens = []
        forward_cells = []

        batch_input_linearity_weight = self.input_linearity_weight
        batch_hidden_linearity_weight = self.hidden_linearity_weight

        if self.training and self.recurrent_dropout_type == 1:
            if self.recurrent_dropout_probability == 0.0:
                raise ValueError("Variational: recurrent_dropout_probability not provided.")

            variational_dropout_mask = get_dropout_mask(
                    self.recurrent_dropout_probability,
                    initial_hidden_state,
            )

        if self.training and self.recurrent_dropout_type == 2:
            if self.recurrent_dropout_probability == 0.0:
                raise ValueError("Dropconnect: recurrent_dropout_probability not provided.")

            dropconnect_input_mask = get_dropout_mask(
                    self.recurrent_dropout_probability,
                    batch_input_linearity_weight,
            )
            dropconnect_hidden_mask = get_dropout_mask(
                    self.recurrent_dropout_probability,
                    batch_hidden_linearity_weight,
            )

            batch_input_linearity_weight = \
                    batch_input_linearity_weight * dropconnect_input_mask
            batch_hidden_linearity_weight = \
                    batch_hidden_linearity_weight * dropconnect_hidden_mask

        init_batch_size = batch_sizes[0] if self.go_forward else batch_sizes[total_timesteps - 1]
        hidden_state = initial_hidden_state.narrow(0, 0, init_batch_size)
        cell_state = initial_cell_state.narrow(0, 0, init_batch_size)

        total_offset = 0
        for ts_offset in range(0, total_timesteps):
            timestep_index = ts_offset if self.go_forward else total_timesteps - 1 - ts_offset
            batch_size = batch_sizes[timestep_index]

            if self.training \
                    and self.truncated_bptt > 0 \
                    and ts_offset > 0 \
                    and ts_offset % self.truncated_bptt == 0:
                hidden_state = hidden_state.detach()
                cell_state = cell_state.detach()

            if self.go_forward:
                if timestep_index > 0:
                    last_batch_size = batch_sizes[timestep_index - 1]
                    dec = last_batch_size - batch_size
                    if dec > 0:
                        forward_hiddens.append(hidden_state.narrow(0, last_batch_size - dec, dec))
                        forward_cells.append(cell_state.narrow(0, last_batch_size - dec, dec))

                        hidden_state = hidden_state.narrow(0, 0, batch_size)
                        cell_state = cell_state.narrow(0, 0, batch_size)

            else:
                if timestep_index < total_timesteps - 1:
                    last_batch_size = batch_sizes[timestep_index + 1]
                    inc = batch_size - last_batch_size
                    if inc > 0:
                        hidden_state = torch.cat(
                                [
                                        hidden_state,
                                        initial_hidden_state.narrow(0, batch_size - inc, inc),
                                ],
                                dim=0,
                        )
                        cell_state = torch.cat(
                                [
                                        cell_state,
                                        initial_cell_state.narrow(0, batch_size - inc, inc),
                                ],
                                dim=0,
                        )

            if self.go_forward:
                input_slice_begin = total_offset
            else:
                input_slice_begin = total_sequences_size - total_offset - batch_size
            cur_input = inputs.narrow(0, input_slice_begin, batch_size)

            proj_input = F.linear(cur_input, batch_input_linearity_weight)
            proj_hidden = F.linear(
                    hidden_state,
                    batch_hidden_linearity_weight,
                    self.hidden_linearity_bias,
            )

            input_gate = torch.sigmoid(
                    proj_input.narrow(1, 0 * self.cell_size, self.cell_size) +
                    proj_hidden.narrow(1, 0 * self.cell_size, self.cell_size))
            forget_gate = torch.sigmoid(
                    proj_input.narrow(1, 1 * self.cell_size, self.cell_size) +
                    proj_hidden.narrow(1, 1 * self.cell_size, self.cell_size))
            cell_tilde = torch.tanh(
                    proj_input.narrow(1, 2 * self.cell_size, self.cell_size) +
                    proj_hidden.narrow(1, 2 * self.cell_size, self.cell_size))
            output_gate = torch.sigmoid(
                    proj_input.narrow(1, 3 * self.cell_size, self.cell_size) +
                    proj_hidden.narrow(1, 3 * self.cell_size, self.cell_size))

            cell_state = input_gate * cell_tilde + forget_gate * cell_state
            if self.cell_clip > 0.0:
                cell_state.clamp_(-self.cell_clip, self.cell_clip)

            hidden_state = F.linear(
                    output_gate * torch.tanh(cell_state),
                    self.proj_linearity_weight,
            )
            if self.proj_clip > 0.0:
                hidden_state.clamp_(-self.proj_clip, self.proj_clip)

            if self.training and self.recurrent_dropout_type == 1:
                hidden_state *= variational_dropout_mask.narrow(0, 0, batch_size)

            total_offset += batch_size

            output_accumulator[timestep_index] = hidden_state

        if self.go_forward:
            forward_hiddens.append(hidden_state)
            hidden_state = torch.cat(forward_hiddens[::-1], dim=0)

            forward_cells.append(cell_state)
            cell_state = torch.cat(forward_cells[::-1], dim=0)

        state_dec = initial_hidden_state.shape[0] - hidden_state.shape[0]
        if state_dec > 0:
            hidden_state = torch.cat(
                    [
                            hidden_state,
                            initial_hidden_state.narrow(0, hidden_state.shape[0], state_dec),
                    ],
                    dim=0,
            )

        state_dec = initial_cell_state.shape[0] - cell_state.shape[0]
        if state_dec > 0:
            cell_state = torch.cat(
                    [
                            cell_state,
                            initial_cell_state.narrow(0, cell_state.shape[0], state_dec),
                    ],
                    dim=0,
            )

        final_state = (hidden_state.unsqueeze(0), cell_state.unsqueeze(0))
        return torch.cat(output_accumulator, dim=0), final_state


class PyUnidirectionalLstm(torch.nn.Module):  # type: ignore

    def __init__(
            self,
            num_layers: int,
            input_size: int,
            hidden_size: int,
            cell_size: int,
            go_forward: bool = True,
            truncated_bptt: int = 0,
            cell_clip: float = 0.0,
            proj_clip: float = 0.0,
            recurrent_dropout_type: int = 0,
            recurrent_dropout_probability: float = 0.0,
            use_skip_connections: bool = False,
    ) -> None:
        super().__init__()

        self.hidden_size = hidden_size
        self.cell_size = cell_size
        self.num_layers = num_layers
        self.use_skip_connections = use_skip_connections

        if num_layers <= 0:
            raise ValueError("num_layers should >= 1.")

        self.layers: List[PyUnidirectionalSingleLayerLstm] = []
        layer_name_prefix = "forward_layer_" if go_forward else "backward_layer_"
        lstm_input_size = input_size
        for layer_idx in range(0, num_layers):
            layer = PyUnidirectionalSingleLayerLstm(
                    lstm_input_size,
                    hidden_size,
                    cell_size,
                    go_forward,
                    truncated_bptt,
                    cell_clip,
                    proj_clip,
                    recurrent_dropout_type,
                    recurrent_dropout_probability,
            )
            self.add_module(
                    layer_name_prefix + str(layer_idx),
                    layer,
            )
            self.layers.append(layer)

            lstm_input_size = hidden_size

    def forward(  # pylint: disable=arguments-differ
            self,
            inputs: torch.Tensor,
            batch_sizes: torch.Tensor,
            initial_state: Tuple[torch.Tensor, torch.Tensor],
    ) -> Tuple[torch.Tensor, Tuple[torch.Tensor, torch.Tensor]]:
        layers_hidden_state, layers_cell_state = initial_state

        output_accumulators = [None] * self.num_layers
        hidden_states = [None] * self.num_layers
        cell_states = [None] * self.num_layers

        layer_inputs = inputs
        for layer_idx in range(0, self.num_layers):
            layer_hidden_state = layers_hidden_state.select(0, layer_idx).unsqueeze(0)
            layer_cell_state = layers_cell_state.select(0, layer_idx).unsqueeze(0)

            output_accumulator, (hidden_state, cell_state) = self.layers[layer_idx](
                    layer_inputs,
                    batch_sizes,
                    (layer_hidden_state, layer_cell_state),
            )
            output_accumulators[layer_idx] = output_accumulator
            hidden_states[layer_idx] = hidden_state
            cell_states[layer_idx] = cell_state

            if layer_idx > 0 and self.use_skip_connections:
                output_accumulator += layer_inputs
            layer_inputs = output_accumulator

        return output_accumulators, (torch.cat(hidden_states, dim=0), torch.cat(cell_states, dim=0))
